crates/sui-types/src/full_checkpoint_content.rs

```diff
@@ -18,11 +21,47 @@ pub struct CheckpointData {
 }
 
 impl CheckpointData {
-    pub fn output_objects(&self) -> Vec<&Object> {
-        self.transactions
+    pub fn latest_live_output_objects(&self) -> Vec<&Object> {
+        let output_objects = self.transactions
             .iter()
             .flat_map(|tx| &tx.output_objects)
-            .collect()
+            .collect::<Vec<_>>();
+        let mut output_objects_by_id = BTreeMap::new();
+        for obj in output_objects {
+            output_objects_by_id
+                .entry(obj.id())
+                .or_insert_with(Vec::new)
+                .push(obj);
+        }
+        let latest_output_objects = output_objects_by_id
+            .into_iter()
+            .flat_map(|(_, objs)| objs.into_iter().max_by_key(|o| o.version()))
+            .collect::<Vec<_>>();
+
+        let removed_object_refs_post_version = self.transactions.iter()
+            .flat_map(|tx| tx.removed_object_refs_post_version()).collect::<Vec<_>>();
+        latest_output_objects.into_iter().filter_map(|obj| {
+            if removed_object_refs_post_version.iter().any(|obj_ref| obj_ref.0 == obj.id() && obj_ref.1 > obj.version()) {
+                None
+            } else {
+                Some(obj)
+            }
+        }).collect()
+    }
+
+    pub fn eventually_removed_object_refs_post_version(&self) -> Vec<ObjectRef> {
```

Comment:
while implementing this, I read the current impl of removed objects codes https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/full_checkpoint_content.rs#L67
do we want to chain unwrapped_then_deleted() there too? @bmwill @lxfind

Reply:
answering my own q, unwrapped_then_deleted is not needed b/c the object does not exist (wrapped) before the txn, thus will not show in the input objects anyways.